FLT/DedekindDomain/FiniteAdeleRing/BaseChange.lean

```diff
@@ -491,42 +495,176 @@ noncomputable def tensorAdicCompletionIntegersTo (v : HeightOneSpectrum A) :
+        . exact (ModuleTopology.continuousAdd _ _).continuous_add
+        intro _ ha _ hb
+        exact add_mem ha hb
+    | tmul b a' =>
+        -- Rewrite tensorAdicCompletionIntegersTo `(b ⊗ₜ a')` to `b • (1 ⊗ₜ a')`
+        simp only [RingHom.coe_range, tensorAdicCompletionIntegersTo, Algebra.comp_ofId,
+          AlgHom.toRingHom_eq_coe, RingHom.coe_coe, Algebra.TensorProduct.lift_tmul,
+          AlgHom.coe_comp, AlgHom.coe_restrictScalars', IsScalarTower.coe_toAlgHom',
+          Function.comp_apply, ValuationSubring.algebraMap_apply,
+          Algebra.TensorProduct.includeRight_apply]
+        -- Now, `f : a' ↦ b • (1 ⊗ₜ a')` is continuous
+        letI : SMul (adicCompletion K v) (L ⊗[K] adicCompletion K v) :=
+            Algebra.TensorProduct.rightAlgebra |>.toSMul
+        let f (y : ↥(adicCompletionIntegers K v)) : (L ⊗[K] adicCompletion K v) :=
+          (Algebra.ofId B (L ⊗[K] adicCompletion K v)) b * (1 : L) ⊗ₜ[K] (y : adicCompletion K v)
+        have hfval : f = fun (y : ↥(adicCompletionIntegers K v)) =>
+              (y : adicCompletion K v) • (Algebra.ofId B (L ⊗[K] adicCompletion K v)) b := by
+          ext y
+          unfold f
+          rw [mul_comm]
+          rfl
+        have hcf : ContinuousAt f a' := by
+          apply Continuous.continuousAt
+          rw [hfval]
+          apply Continuous.smul continuous_subtype_val continuous_const
+        -- So, because `A` is dense in `A'`, `b • (1 ⊗ₜ a') ∈ closure f '' A`
+        have hy : a' ∈ closure (Set.range (algebraMap A _)) := by
+          apply IsDedekindDomain.denseRange_of_integerAlgebraMap
+        apply mem_closure_image hcf hy
+        constructor
+        . exact isClosed_closure
+        -- Finally, `b • (1 ⊗ₜ a) = (b * a) • (1 ⊗ₜ 1)`, so `f '' A ⊆ algebraMap '' B`
+        rintro u ⟨_, ⟨a, rfl⟩, rfl⟩
+        apply subset_closure
+        use algebraMap A B a * b
+        unfold f
+        rw [Algebra.algebraMap_eq_smul_one (A := (adicCompletionIntegers K v)) a,
+          coe_smul_adicCompletionIntegers, ← TensorProduct.smul_tmul, Algebra.ofId_apply,
+          Algebra.TensorProduct.algebraMap_apply, RingHom.map_mul, ← Algebra.smul_def]
+        simp
+
+open TensorProduct.AlgebraTensorModule in
+open scoped Classical in
+attribute [local instance] Algebra.TensorProduct.rightAlgebra in
+omit [Algebra.IsSeparable K L] [IsDomain B] [Algebra.IsIntegral A B]
+    [Module.Finite A B] [IsDedekindDomain B] [IsFractionRing B L]  in
+lemma tensorAdicCompletionIsClosedRange :
+    IsClosed (SetLike.coe (tensorAdicCompletionIntegersTo A K L B v).range) := by
+  -- `B ⊗[A] 𝒪_v` is a subgroup of `L ⊗[K] K_v`, so we can show it's closed
+  -- by showing that it's open.
+  rw [← Subalgebra.coe_toSubring, ← Subring.coe_toAddSubgroup]
+  letI : Module (adicCompletion K v) (L ⊗[K] adicCompletion K v) :=
+      Algebra.TensorProduct.rightAlgebra.toModule
+  letI : SMul (adicCompletion K v) (L ⊗[K] adicCompletion K v) :=
+      Algebra.TensorProduct.rightAlgebra.toSMul
```

Comment:
You don't need these, typeclass inference will take care of them because of the local instance on line 560.